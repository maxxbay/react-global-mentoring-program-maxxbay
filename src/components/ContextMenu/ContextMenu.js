import React from 'react';
import './ContextMenu.scss';
import PropTypes from 'prop-types';

const ContextMenu = ({ show, onClose, onEdit, onDelete }) => {
  const handleEdit = e => {
    onEdit();
    onClose(e);
  };

  const handleDelete = e => {
    onDelete();
    onClose(e);
  };

  return (
    show && (
      <div className="context-menu">
        <button className="context-menu__item" onClick={handleEdit}>
          Edit
        </button>
        <button className="context-menu__item" onClick={handleDelete}>
          Delete
        </button>
      </div>
    )
  );
};
ContextMenu.propTypes = {
  show: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  onEdit: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired,
};
export default ContextMenu;
