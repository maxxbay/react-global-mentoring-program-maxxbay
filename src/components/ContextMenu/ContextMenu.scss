.context-menu {
    position: absolute;
    background-color: #424242;
    border-radius: 4px;
    box-shadow: 0 2px 8px rgba(0, 0, 0, 0.3);
    overflow: hidden;

    &__item {
        display: block;
        padding: 12px 24px;
        cursor: pointer;
        text-align: left;
        width: 100%;
        border: none;
        background-color: #424242;
        color: #ffffff;
        font-size: 16px;
        line-height: 1.5;

        &:hover {
            background-color: #616161;
        }
    }
}