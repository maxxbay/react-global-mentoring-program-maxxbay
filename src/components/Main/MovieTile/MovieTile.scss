.movie-tile {
    position: relative;
    display: flex;
    flex-direction: column;
    padding: 10px;
    border-radius: 5px;
    background-color: #424242;
    cursor: pointer;
    width: calc(33.33% - 20px);
    margin: 0 10px 20px 0;

    @media screen and (max-width: 1024px) {
        width: calc(50% - 20px);
    }

    @media screen and (max-width: 767px) {
        width: 100%;
    }

    &:nth-child(3n) {
        margin-right: 0;
    }
         &__context-menu-wrapper {
             position: absolute;
             top: 0;
             right: 0;
             display: flex;
             flex-direction: column;
             align-items: flex-end;
         }

    &__image {
        width: 100%;
        height: 300px;
        object-fit: contain;
        border-radius: 5px;
    }

    &__info {
        margin-top: 10px;
        display: flex;
        flex-direction: column;
    }

    &__title {
        font-size: 18px;
        font-weight: 500;
        color: #fff;
        margin-bottom: 5px;
    }

    &__year,
    &__genres {
        font-size: 14px;
        color: #808080;
        margin-bottom: 5px;
    }

    &__rating {
        font-size: 14px;
        color: #808080;
        margin-bottom: 5px;
    }

    &__duration {
        font-size: 14px;
        color: #808080;
        margin-bottom: 5px;
    }

    &__description {
        font-size: 14px;
        color: #808080;
        text-align: justify;
    }

    &__context-menu-toggle {
        position: absolute;
        top: 5px;
        right: 5px;
        font-size: 16px;
        color: #fff;
        background-color: transparent;
        border: none;
        cursor: pointer;
    }

    &__context-menu {
        position: absolute;
        top: 25px;
        right: 5px;
        display: flex;
        flex-direction: column;
        background-color: #424242;
        border-radius: 5px;
        z-index: 10;

        &-item {
            font-size: 14px;
            color: #fff;
            background-color: transparent;
            border: none;
            padding: 5px 10px;
            cursor: pointer;

            &:hover {
                background-color: #666;
            }
        }
        
    }
}
