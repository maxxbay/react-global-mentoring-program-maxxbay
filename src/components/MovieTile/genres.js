export const genres = [
  'Action',
  'Animation',
  'Adventure',
  'Comedy',
  'Drama',
  'Horror',
  'Family',
  'Science Fiction',
  'Thriller',
  'Fantasy',
];
