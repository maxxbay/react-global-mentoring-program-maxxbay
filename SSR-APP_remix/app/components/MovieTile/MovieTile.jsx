import React, { useState, useEffect } from 'react';
import { Link, useNavigate } from 'react-router-dom';
import PropTypes from 'prop-types';
import ContextMenu from '../ContextMenu/ContextMenu';

import stylesUrl from '~/styles/MovieTile.css';

export function MovieTileLinks() {
  return <link rel="stylesheet" href={stylesUrl} />;
}

const MovieTile = ({ movie, onClick }) => {
  const navigate = useNavigate();

  const { title, release_date, genres, poster_path } = movie;
  const [showContextMenu, setShowContextMenu] = useState(false);

  const toggleContextMenu = e => {
    e.stopPropagation();
    setShowContextMenu(prevShowContextMenu => !prevShowContextMenu);
  };

  const handleEdit = () => {
    navigate(`/movies/${movie.id}/edit`, { state: { movie } });
  };

  const handleDelete = () => {
    console.log('Delete action for movie:', movie);
  };

  useEffect(() => {
    const handleClickOutside = e => {
      if (showContextMenu) {
        setShowContextMenu(false);
      }
    };
    document.addEventListener('click', handleClickOutside);
    return () => {
      document.removeEventListener('click', handleClickOutside);
    };
  }, [showContextMenu]);

  return (
    <Link to={`/movies/${movie.id}`} className="movie-tile">
      <button className="movie-tile__button">
        <img
          className="movie-tile__image"
          src={`https://image.tmdb.org/t/p/w185${poster_path}`}
          alt={title}
        />
        <div className="movie-tile__info">
          <h2 className="movie-tile__title">{title}</h2>
          <p className="movie-tile__year">{release_date}</p>
          <p className="movie-tile__genres">
            {genres ? genres.join(', ') : ''}
          </p>
          <p className="movie-tile__rating">Rating: {movie.vote_average}</p>
          <p className="movie-tile__duration">Duration: {movie.runtime} min</p>
          <p className="movie-tile__description">
            {movie.overview || 'No overview available.'}
          </p>
        </div>
      </button>

      <div className="movie-tile__context-menu-wrapper">
        <button
          className="movie-tile__context-menu-toggle"
          aria-label="context menu"
          onClick={toggleContextMenu}
        >
          &#x2026;
        </button>
        <ContextMenu
          show={showContextMenu}
          onClose={toggleContextMenu}
          onEdit={handleEdit}
          onDelete={handleDelete}
        />
      </div>
    </Link>
  );
};
MovieTile.propTypes = {
  movie: PropTypes.shape({
    title: PropTypes.string.isRequired,
    id: PropTypes.number.isRequired,
    release_date: PropTypes.string.isRequired,
    genres: PropTypes.arrayOf(PropTypes.string).isRequired,
    poster_path: PropTypes.string.isRequired,
    vote_average: PropTypes.number.isRequired,
    runtime: PropTypes.number,
    overview: PropTypes.string,
  }).isRequired,
  onClick: PropTypes.func,
};

export default MovieTile;
