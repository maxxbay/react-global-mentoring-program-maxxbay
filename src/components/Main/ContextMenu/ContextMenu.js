import React from 'react';
import './ContextMenu.scss';

const ContextMenu = ({ show, onClose, onEdit, onDelete }) => {
  const handleEdit = e => {
    e.stopPropagation();
    onEdit();
    onClose();
  };

  const handleDelete = e => {
    e.stopPropagation();
    onDelete();
    onClose();
  };

  return (
    show && (
      <div className="context-menu">
        <button className="context-menu__item" onClick={handleEdit}>
          Edit
        </button>
        <button className="context-menu__item" onClick={handleDelete}>
          Delete
        </button>
      </div>
    )
  );
};

export default ContextMenu;
