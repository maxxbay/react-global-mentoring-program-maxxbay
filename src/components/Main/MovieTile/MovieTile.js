import React, { useState } from 'react';
import './MovieTile.scss';

const MovieTile = ({ movie, onClick, onEdit, onDelete }) => {
  const { title, release_date, genres, poster_path } = movie;
  const [showContextMenu, setShowContextMenu] = useState(false);

  const toggleContextMenu = () => {
    setShowContextMenu(!showContextMenu);
  };

  const handleTileClick = () => {
    if (typeof onClick === 'function') {
      onClick(movie);
    }
  };

  const handleEdit = () => {
    if (typeof onEdit === 'function') {
      onEdit(movie);
    }
  };

  const handleDelete = () => {
    if (typeof onDelete === 'function') {
      onDelete(movie);
    }
  };

  return (
    <div className="movie-tile" onClick={handleTileClick}>
      <img
        className="movie-tile__image"
        src={`https://image.tmdb.org/t/p/w185${poster_path}`}
        alt={title}
      />
      <div className="movie-tile__info">
        <h2 className="movie-tile__title">{title}</h2>
        <p className="movie-tile__year">{release_date}</p>
        <p className="movie-tile__genres">{genres.join(', ')}</p>
        <p className="movie-tile__rating">Rating: {movie.vote_average}</p>
        <p className="movie-tile__duration">Duration: {movie.runtime} min</p>
        <p className="movie-tile__description">{movie.overview}</p>
      </div>
      <button
        className="movie-tile__context-menu-toggle"
        onClick={toggleContextMenu}
      >
        &#x2026;
      </button>
      {showContextMenu && (
        <div className="movie-tile__context-menu">
          <button
            className="movie-tile__context-menu-item"
            onClick={handleEdit}
          >
            Edit
          </button>
          <button
            className="movie-tile__context-menu-item"
            onClick={handleDelete}
          >
            Delete
          </button>
        </div>
      )}
    </div>
  );
};

export default MovieTile;
