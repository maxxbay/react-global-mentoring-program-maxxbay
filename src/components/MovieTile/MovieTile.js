import React, { useState, useEffect } from 'react';
import './MovieTile.scss';
import ContextMenu from '../ContextMenu/ContextMenu';

const MovieTile = ({ movie, onClick }) => {
  const { title, release_date, genres, poster_path } = movie;
  const [showContextMenu, setShowContextMenu] = useState(false);

  const toggleContextMenu = e => {
    e.stopPropagation();
    setShowContextMenu(prevShowContextMenu => !prevShowContextMenu);
  };

  const handleEdit = () => {
    console.log('Edit action for movie:', movie);
  };

  const handleDelete = () => {
    console.log('Delete action for movie:', movie);
  };

  const handleTileClick = () => {
    if (typeof onClick === 'function') {
      onClick(movie);
    }
  };

  useEffect(() => {
    const handleClickOutside = e => {
      if (showContextMenu) {
        setShowContextMenu(false);
      }
    };
    document.addEventListener('click', handleClickOutside);
    return () => {
      document.removeEventListener('click', handleClickOutside);
    };
  }, [showContextMenu]);

  return (
    <div className="movie-tile" onClick={handleTileClick}>
      <img
        className="movie-tile__image"
        src={`https://image.tmdb.org/t/p/w185${poster_path}`}
        alt={title}
      />
      <div className="movie-tile__info">
        <h2 className="movie-tile__title">{title}</h2>
        <p className="movie-tile__year">{release_date}</p>
        <p className="movie-tile__genres">{genres.join(', ')}</p>
        <p className="movie-tile__rating">Rating: {movie.vote_average}</p>
        <p className="movie-tile__duration">Duration: {movie.runtime} min</p>
        <p className="movie-tile__description">{movie.overview}</p>
      </div>
      <div className="movie-tile__context-menu-wrapper">
        <button
          className="movie-tile__context-menu-toggle"
          aria-label="context menu"
          onClick={toggleContextMenu}
        >
          &#x2026;
        </button>
        <ContextMenu
          show={showContextMenu}
          onClose={toggleContextMenu}
          onEdit={handleEdit}
          onDelete={handleDelete}
        />
      </div>
    </div>
  );
};

export default MovieTile;
