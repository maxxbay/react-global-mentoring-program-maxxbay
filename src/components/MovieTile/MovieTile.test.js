import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import MovieTile from './MovieTile';
import ContextMenu from '../ContextMenu/ContextMenu';
import '@testing-library/jest-dom';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => jest.fn(),
}));

const mockMovie = {
  id: 1,
  title: 'Mock Movie',
  release_date: '2022-01-01',
  genres: ['Action', 'Adventure'],
  poster_path: '/mock-poster.jpg',
  vote_average: 7.5,
  runtime: 120,
  overview: 'This is a mock movie.',
};

describe('MovieTile component', () => {
  it('renders correctly', () => {
    const { container } = render(<MovieTile movie={mockMovie} />);
    expect(container).toMatchSnapshot();
  });

  it('renders movie data correctly', () => {
    render(<MovieTile movie={mockMovie} />);
    expect(screen.getByText(mockMovie.title)).toBeInTheDocument();
    expect(screen.getByText(mockMovie.release_date)).toBeInTheDocument();
    expect(screen.getByText(mockMovie.genres.join(', '))).toBeInTheDocument();
    expect(
      screen.getByText(`Rating: ${mockMovie.vote_average}`)
    ).toBeInTheDocument();
    expect(
      screen.getByText(`Duration: ${mockMovie.runtime} min`)
    ).toBeInTheDocument();
    expect(screen.getByText(mockMovie.overview)).toBeInTheDocument();
    expect(screen.getByAltText(mockMovie.title)).toHaveAttribute(
      'src',
      `https://image.tmdb.org/t/p/w185${mockMovie.poster_path}`
    );
  });

  it('calls onClick function when clicked', () => {
    const handleClick = jest.fn();
    render(<MovieTile movie={mockMovie} onClick={handleClick} />);
    fireEvent.click(screen.getByText(mockMovie.title));
    expect(handleClick).toHaveBeenCalledWith(mockMovie);
  });

  it('opens context menu when toggle button is clicked', () => {
    render(<MovieTile movie={mockMovie} />);
    const toggleButton = screen.getByRole('button', { name: 'context menu' });
    fireEvent.click(toggleButton);
    expect(screen.getByText('Edit')).toBeInTheDocument();
    expect(screen.getByText('Delete')).toBeInTheDocument();
  });

  it('does not open context menu when toggle button is not clicked', () => {
    render(<MovieTile movie={mockMovie} />);
    expect(screen.queryByText('Edit')).not.toBeInTheDocument();
    expect(screen.queryByText('Delete')).not.toBeInTheDocument();
  });
});

describe('ContextMenu component', () => {
  it('renders correctly when show is true', () => {
    const { asFragment } = render(
      <ContextMenu
        show
        onClose={jest.fn()}
        onEdit={jest.fn()}
        onDelete={jest.fn()}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('does not render when show is false', () => {
    const onClose = jest.fn();

    render(
      <ContextMenu
        show={false}
        onClose={onClose}
        onEdit={jest.fn()}
        onDelete={jest.fn()}
      />
    );

    expect(screen.queryByText('Edit')).not.toBeInTheDocument();
    expect(screen.queryByText('Delete')).not.toBeInTheDocument();
  });

  it('calls onClose function when edit button is clicked', () => {
    const handleClose = jest.fn();
    render(
      <ContextMenu
        show
        onClose={handleClose}
        onEdit={jest.fn()}
        onDelete={jest.fn()}
      />
    );
    const editButton = screen.getByText('Edit');
    fireEvent.click(editButton);
    expect(handleClose).toHaveBeenCalled();
  });

  it('calls onClose function when delete button is clicked', () => {
    const handleClose = jest.fn();
    render(
      <ContextMenu
        show
        onClose={handleClose}
        onEdit={jest.fn()}
        onDelete={jest.fn()}
      />
    );
    const deleteButton = screen.getByText('Delete');
    fireEvent.click(deleteButton);
    expect(handleClose).toHaveBeenCalled();
  });
});
