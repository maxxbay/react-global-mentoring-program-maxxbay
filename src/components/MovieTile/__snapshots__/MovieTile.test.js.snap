// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextMenu component renders correctly when show is true 1`] = `
<div>
  <div
    class="context-menu"
  >
    <button
      class="context-menu__item"
    >
      Edit
    </button>
    <button
      class="context-menu__item"
    >
      Delete
    </button>
  </div>
</div>
`;

exports[`MovieTile component renders correctly 1`] = `
<div>
  <div
    class="movie-tile"
  >
    <img
      alt="Mock Movie"
      class="movie-tile__image"
      src="https://image.tmdb.org/t/p/w185/mock-poster.jpg"
    />
    <div
      class="movie-tile__info"
    >
      <h2
        class="movie-tile__title"
      >
        Mock Movie
      </h2>
      <p
        class="movie-tile__year"
      >
        2022-01-01
      </p>
      <p
        class="movie-tile__genres"
      >
        Action, Adventure
      </p>
      <p
        class="movie-tile__rating"
      >
        Rating: 
        7.5
      </p>
      <p
        class="movie-tile__duration"
      >
        Duration: 
        120
         min
      </p>
      <p
        class="movie-tile__description"
      >
        This is a mock movie.
      </p>
    </div>
    <div
      class="movie-tile__context-menu-wrapper"
    >
      <button
        aria-label="context menu"
        class="movie-tile__context-menu-toggle"
      >
        â€¦
      </button>
    </div>
  </div>
</div>
`;
